File: spark-scala/src/main/java/org/guoqing/sparkscala/controller/Count.scala
package org.guoqing.sparkscala.controller

import org.apache.spark.{SparkConf, SparkContext}
import org.guoqing.sparkscala.config.InitSpark

/**
  * ${DESCRIPTION}
  *
  * @author dengguoqing
  * @since 2018-11-06
  */
class Count {
  def wordCount(inputFile: String, outputFile: String) = {
    /*val conf = new SparkConf().setMaster("local").setAppName("wordCount")
    val sc = new SparkContext(conf)*/

    val sc = InitSpark.sparkConf()
    val input = sc.textFile(inputFile)
    val words = input.flatMap(line => line.split(" "))
    val counts = words.map(word => (word, 1)).reduceByKey { case (x, y) => x + y }
    counts.saveAsTextFile(outputFile)
  }
}

File: spark-java/src/main/java/org/guoqing/sparkjava/controller/Count.java
package org.guoqing.sparkjava.controller;

import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.guoqing.sparkjava.config.InitSpark;
import scala.Tuple2;

import java.util.Arrays;

/**
 * ${DESCRIPTION}
 *
 * @author dengguoqing
 * @since 2018-11-06
 */
public class Count {

    /**
     * 统计输入文件的单词个数，并输出到指定文件
     *
     * @param inputFile  输入文件位置
     * @param outputFile 输出文件位置
     */
    public void wordCount(String inputFile, String outputFile) {
        JavaSparkContext sc = InitSpark.sparkContext();
        JavaRDD<String> input = sc.textFile(inputFile);
        JavaRDD<String> words =
                input.flatMap(x -> Arrays.asList(x.split(" ")).iterator());
        JavaPairRDD<String, Integer> counts =
                words.mapToPair(x -> new Tuple2<>(x, 1))
                        .reduceByKey((x, y) -> (x + y));
        counts.saveAsTextFile(outputFile);
    }
}
